text/0000-light-dom.md

```diff
@@ -0,0 +1,293 @@
+
+```js
+import { LightningElement } from "lwc";
+
+// Example of a shadow DOM component
+class ShadowDOMComponent extends LightningElement {}
+
+// Example of a light DOM component
+class LightDOMComponent extends LightningElement {
+  static shadowDOM = false;
+}
+
+// Default value
+console.log(LightningElement.shadowDOM); // true
+```
+
+The `shadowDOM` property is looked up by the LWC engine when the component is instantiated for the first time and is cached for future instantiation. Changing the value of the `shadowDOM` static property after the first instantiation doesn't influence whether components are rendered in the light DOM or in the shadow DOM.
+
+It also means that developers should be careful not to override the `shadowDOM` static property value when inheriting from another component. Switching a component mode from shadow DOM to light DOM (and vice-versa) in the child class would certainly break logic in the base class.
+
+```js
+import { LightningElement } from "lwc";
+
+class Base extends LightningElement {}
+
+class Child extends Base {
+  // ⚠️ Changing the shadowDOM static property value in a component class inheriting from a base
+  // will certainly class some unexpected issue.
+  static shadowDOM = false;
+}
+```
+
+### Component features when using Light DOM
+
+Some of the LWC component capabilities are directly inherited from Shadow DOM, or emulated by the synthetic-shadow. Despite the use of Light DOM, we’d like to keep these features available, even if their behavior is slightly adapted to the Light DOM:
+
+#### Slots
+
+As mentioned before, the component composition model in LWC is provided by slots. Light DOM will provide the same mental model for developers building Light DOM components.
+
+In Light DOM, `<slot>` will denote the place where the slotted component will be attached. The `<slot>` element itself won't be rendered. The slotted content (or the fallback content) will be flattened to the parent element at runtime.
+
+Since the `<slot>` element itself isn't rendered, adding attributes or event listeners to the `<slot>` element in the template will throw a compiler error.
+
+Due to the above differences from regular `<slot>`, LWC compiler will enforce the presence of `<slot lwc:no-shadow>` on Light DOM slots to make it explicit to the user that these are different slots.
+
+##### Timing
+
+In terms of timing, slots in light DOM will behave similarly to the current synthetic shadow slots. For example:
+
+```html
+<!-- x/slottable.html -->
+<template lwc:no-shadow>
+  <slot lwc:no-shadow name="foo"></slot>
+  <slot lwc:no-shadow></slot>
+  <slot lwc:no-shadow name="bar"></slot>
+</template>
+```
+
+```html
+<!-- app.html -->
+<template>
+  <x-slottable>
+    <x-baz></x-baz>
+    <x-foo slot="foo"></x-foo>
```

Comment:
**Side note:**

On the consumer side in the `app.html`, the example below is using the standard `slot` attribute. From the compiler standpoint, it is impossible to know if the `<x-slottable>` uses the light DOM or the shadow DOM slotting algorithm.

I was under the original impression that we could also ask developers to use a custom directive to indicate to the template compiler to use the light DOM slotting algorithm to avoid this branching at runtime. For example using `slot:no-shadow` instead of `slot`. That said I came to the conclusion that it is not possible.

While this custom directive works well with named slots I don't see how this would work with the default slot. In the example below I don't see how it is possible to indicate that the text content should go to the light DOM instead of the shadow DOM.

```html
<template>
  <x-slottable>
    Some content
  </x-slottable>
</template>
```

Reply:
This is a good point. I don't know how to resolve it, so perhaps slots should be moved to a separate RFC entirely.

Reply:
After discussing this in a meeting, it seems the only way to resolve the above example is to move the logic to the runtime. I.e. we have to figure out at runtime rather than compile time whether a slot is shadow or light.

The only workaround I can think of would be to make a hard requirement on light DOM slots that they have some top-level element that we can mark:

```html
<template>
  <x-slottable>
    <div lwc:slot> <!-- required -->
      Some content
    </div>
  </x-slottable>
</template>
```

This isn't great for developer ergonomics, but it would allow us to figure this out at compile time.

Reply:
We decided that whether slots are light or shadow will be determined at runtime.